ptclient/ldap.c

```diff
@@ -1387,11 +1386,29 @@ static int ptsmodule_make_authstate_group(
         goto done;
     }
 
-    *dsize = sizeof(struct auth_state);
+    // there must be one
+    entry = ldap_first_entry(ptsm->ld, res);
+    int numvals = 0;
+    if (ptsm->groupmember_attribute) {
+        vals = ldap_get_values(ptsm->ld, entry, (char *)ptsm->groupmember_attribute);
+        numvals = ldap_count_values(vals);
+    }
+
+    // now fetch the list of members!
+
+    *dsize = sizeof(struct auth_state) + (numvals * sizeof(struct auth_ident));
     *newstate = xzmalloc(*dsize);
+    (*newstate)->ngroups = numvals;
     strcpy((*newstate)->userid.id, canon_id);
     (*newstate)->userid.hash = strhash(canon_id);
     (*newstate)->mark = time(0);
+    int i;
+    for (i = 0; i < numvals; i++) {
+        char **rdn = ldap_explode_rdn(vals[i], 1);
+        strlcat((*newstate)->groups[i].id, rdn[0], sizeof((*newstate)->groups[i].id));
```

Comment:
This is looking better but it still feels non-LDAPpy.  Looks like it's assuming the first component of the DN is a usable unique key?  But the whole point of a dn is that _it_ is the unique key, and so its components don't have to be.

In my experience, real world user dns are things like `cn=Ellie Timoney,ou=Cyrus Developers,o=Fastmail,c=au`.  I'm under the impression that the schemas I'm familiar with were pretty standard at the time, but I don't know that with certainty, and even if they were, I don't know if they still are.  In the faked up directory.ldif data, the first dn component is the uid, and so it is coincidentally unique in the way you're trying to use it, but that's mostly because I didn't feel like making up common names or a complicated schema for the entries I was creating.  That's my bad I guess, cause in taking that shortcut I've implied a simplicity that doesn't necessarily exist.  

I think the goal here is to get a list of uids back.  To turn a list of dns into a list of uid's, you would do a scope=base search (which returns exactly 1 result or a no-such-object error) for each of those dn's, fetching just the uid attribute from each.  I think our fakeldapd should be able to handle these requests already.  The attribute to fetch should be configurable, I think, because it might not always be called "uid" (e.g. it might be called "mail").  This might be what `ldap_user_attribute` is for, in which case it can be re-used.  If not, I think it's fine to hard-code "uid" for now, and someone who uses something else in the real world can tell us all about it so we can fix it properly. :)

Interestingly, expanding the group in this way will also mean that if the group contains dangling member entries for people that are no longer in the organisation, those dn searches will return zero results, and the final result will be automatically filtered to only contain uids that really exist -- nice!

Reply:
Yeah - I've done exactly that!  It still passes tests :)